design-proposals/klip-58-limit-clause-for-pull-queries.md

```diff
@@ -0,0 +1,217 @@
+- When a pull query is executed over a stream the `LIMIT` clause should output results from the beginning of the stream.
+
+## What is not in scope
+- Pull queries over `TABLES` don't offer any ordering guarantees today, so the exact subset of data that will returned 
+by restricting the output using `LIMIT` won't be covered in this klip.
+- Performance profile of the `LIMIT` clause (?)
+
+## Value/Return
+
+The ability to `LIMIT` should help users prototyping in ksqlDB with pull by giving them access to a subset of the results 
+that they would expect to see without the `LIMIT` in place. ksqlDB users will not have to issue exepnsive table scans to
+just get a quick look of what the underlying table or stream looks like (the schema and the rows/records).
+
+## Public APIS
+
+#####ksqlDB query language: 
+Users will be able to add a `LIMIT n` clause at the end of their pull queries to retrievea subset of the data from tables 
+and streams. Earlier, users would get this error if they issued a pull query with a `LIMIT` clause: 
+```
+Pull queries don't support LIMIT clauses. See https://cnfl.io/queries for more info. 
+Add EMIT CHANGES if you intended to issue a push query.
+```
+
+Syntax evolution:
+```roomsql
+SELECT select_expr [, ...]                 SELECT select_expr [, ...] 
+  FROM table/stream                          FROM table/stream        
+  [ WHERE where_condition ]    ────────►     [ WHERE where_condition ]
+  [ AND window_bounds ];                     [ AND window_bounds ]    
+                                           + [ LIMIT num_records ];  
+```                                     
+                                               
+
+## Design
+### For Pull Queries over Tables
+Pull query execution relies on building a logical plan which is then translated into a physical plan.The `LIMIT` clause 
+of pull queries will be translated into a `QueryLimitNode` in the logical plan. If present, the `QueryLimitNode` will be 
+the parent node of `QueryProjectNode` in the logical plan. The `QueryLimitNode` will then be translated into an operator 
+called the `LimitOperator` in the physical plan by the pull query physical plan builder. The `LimitOperator` will wil be 
+the parent operator of `ProjectOperator` in the physical plan. The `QueryLimitNode` and `LimitOperator` will store the 
+number of rows to return in them.
+
+Functionally, the `LimitOperator` will stop returning rows flowing down from upstream operators in the physical plan 
+topology by keeping a counter of the number of rows returned so far and stop scanning further once the limit is reached. 
+This will result in returning the first set of n (where n is the limit requested in the query) rows that it gets from 
+upstream operators and stop scanning for rows after that. If the limit isn't reached and we run out of rows to return 
+(exhausting the rows that conform to our query), then the operator will just return. 
```

Comment:
I think the existing design with be able to more seamlessly work with `ORDER BY` so I'm inclined to keep it as is, especially since it's simpler.

Generally for the motivation section above (testing out queries) I don't think we should expect very high values of `n`

Reply:
Works for me. Just wanted to point it out, that it might break for larger `n` (I agree it might not be an issue in practice...)

Reply:
yes, @vvcephei . That's the plan